### ISSUE TITLE
[MiniViT] hyperparameters setting of CIFAR-100 and CIFAR-10

### ISSUE THREAD

**Issue by shadowpa0327** [microsoft/Cream#83]
Hi, could you please provide the detailed training setting to rebuild the result on CIFAR-100 and CIFAR-10 ? 
Thank you !

**Comment by Weigaa** [comment]
It seems that miniViT achieves 99.3% acc in cifar-10, I also want to retrain the model to get the same acc in the paper.

**Comment by wkcn** [comment]
@shadowpa0327 Sorry for my reply late.
@shadowpa0327 @Weigaa Thanks for your attention to our work!

We finetune Mini-DeiT-Base-384 on CIFAR10 and CIFAR100 under 384x384 image resolution with the same hyper-parameters.
The pretrained model can be downloaded in the [link](https://github.com/DominickZhang/MiniViT-model-zoo/releases/download/v1.0.0/mini_deit_base_patch16_384.pth).

The hyper-parameters are the following:
an SGD optimizer is adopted with learning rate 0.005 , batch size 32*8 (namely 256), weight decay 10−4 , and disabled random erase.
We train models for 1000 epochs.

Note that 0.005 is the linear scaled learning rate, which is equal to lr (1e-2) * 32 * 8 / 512 = 0.005.

The training command:
`python -m torch.distributed.launch --nproc_per_node=8 --use_env main.py --model mini_deit_base_patch16_384 --batch-size 32 --data-path ./cifar100 --output_dir ./outputs --finetune ./checkpoints/mini_deit_base_patch16_384.pth  --reprob 0 --opt sgd  --data-set CIFAR --dist-eval  --input-size 384 --lr 1e-2  --weight-decay 1e-4 --epochs 1000`

**Comment by Weigaa** [comment]
@wkcn Thanks a lot, I'll try it soon. And If I need to change the lr to 0.000625 and increase the epochs to 8000 due to I just have one GPU? and I need to download cifar-10/cifar-100 to ./cifar10 or ./cifar100 before training or it can be downloaded automatically?

**Comment by shadowpa0327** [comment]
@wkcn  Thanks a lot for the reply, and thanks again for the great work !

**Comment by wkcn** [comment]
> @wkcn Thanks a lot for the reply, and thanks again for the great work !

You are welcome !

**Comment by Weigaa** [comment]
@wkcn Thanks again. Indeed, I had some problems with my previous idea. When I use a smaller batch size, I really should spend less epoch. In addition, do I need to halve my learning rate because my BSZ has been halved? (due to I need to do more gradient updates for one epoch)

**Comment by wkcn** [comment]
> @wkcn Thanks again. Indeed, I had some problems with my previous idea. When I use a smaller batch size, I really should spend less epoch. In addition, do I need to halve my learning rate because my BSZ has been halved? (due to I need to do more gradient updates for one epoch)

@Weigaa You are welcome : ) Take an example:
If the total batch size is 256, the number of iterations in an epoch is 400. The number of epochs is 1000. The linear scaled learning rate is 0.005. (lr (1e-2) * 32 * 8 / 512 = 0.005.)

When the total batch size is halved to 128, the number of iterations in an epoch is doubled to 800.  The number of epochs is still 1000. The linear scaled learning rate is 0.0025 (lr (1e-2) * 128 / 512 = 0.0025). You do not need to change the arguments (--lr 1e-2 --epoch 1000).

**Comment by Weigaa** [comment]
@wkcn 感谢，我注意到你们通过对学习率进行简单的线性缩放完成了针对bsz和num_gpus的自动修改，这确实很方便！我注意到finetune似乎非常耗时，我在单台P100上finetune 1个epoch就需要花费超过3个小时的时间。有什么方法可以加快这个过程吗？另外，我对本文提出的3M小模型更感兴趣，请问可以提供关于它的finetune策略以及checkpoint嘛？
很抱歉提出这么多问题打扰您的正常工作，希望能得到您的回复。

**Comment by wkcn** [comment]
@Weigaa Mini-DeiT-Base模型虽然参数量减少了，但计算量仍然是base级别，而且输入分辨率是384x384, 计算量很大，因此很耗时。抱歉没有加速的方法。

可以尝试更小级别的模型，比如Tiny或Small, 用224分辨率训练。

我们没有做Mini-DeiT-Base-384以外的模型的下游任务实验。你可以用同样的超参数(`--reprob 0 --opt sgd --lr 1e-2 --weight-decay 1e-4 --epochs 1000`) 在CIFAR数据集上finetune包括3M规模等其他小模型。训练到100多个epoch就能得到比较好的结果了。

欢迎提问 : )

**Comment by Weigaa** [comment]
@wkcn 请问有用于tiny和small模型finetune的pretrain模型checkpoint嘛？

**Comment by wkcn** [comment]
@Weigaa 
在Model Zoo里，我们提供了Tiny/Small/Base的224分辨率模型

https://github.com/microsoft/Cream/tree/main/MiniViT/Mini-DeiT#model-zoo

**Comment by Weigaa** [comment]
@wkcn 我们在P100 GPU安装rpe_ops时会遇到错误，导致无法安装，具体错误截图如下：
![OMOVLUF8ATP7EBSNHY_$L8U](https://user-images.githubusercontent.com/38322329/168230590-738e3385-53b1-46da-a96e-a1d63197cc06.png)
![0XGUOJ%RE4V5JTHMPMV$8J6](https://user-images.githubusercontent.com/38322329/168230687-041aba2f-513f-4376-a4b0-2615622b5377.png)
我们使用的Pytorch版本是1.10.0，Cuda版本是11.3，这种配置在V100 GPU上编译rpe_ops时不会出错，但是出现在了P100，请问有什么该问题的解决方案吗？

**Comment by wkcn** [comment]
@Weigaa 可能是P100机器装的gcc的版本比较低，可以对比一下V100机器和P100机器的gcc版本。

如果两台机器的CUDA大版本一致（都是11），可以把V100上编译好的*.so文件复制过来，P100机器应该也能使用。
iRPE也可以不编译C++库的。

**Comment by Weigaa** [comment]
@wkcn 感谢您的回复。确实是因为P100上的gcc版本过低（5.5.0）而V100上是9.4.0导致的，我将P100上的gcc版本提升到与V100一致之后，程序就可以正常安装了，后续运行也没有问题。

**Comment by Weigaa** [comment]
@wkcn 您好，我们在使用单节点小于128的batchsize数的时候，Cifar-10和Cifar-100的finetune均没有问题，但在执行bsz为256时，有以下错误提示，请问您对如何解决该问题有哪些建议吗？
![8bed019663341a9d62ca022f81fcc6b](https://user-images.githubusercontent.com/38322329/169361661-d068641f-7a20-4292-bd68-3a63d87322bd.png)
![0e95ae403a767ebcc08ac5cfc6385ef](https://user-images.githubusercontent.com/38322329/169361693-6396e875-398d-450b-8d59-009ab20d82ba.png)
我们用于微调cifar-10的指令如下：
python -m torch.distributed.launch --nproc_per_node=1 --use_env main.py --model mini_deit_tiny_pa
tch16_224 --batch-size 256 --data-path ./cifar10 --output_dir ./outputs --finetune mini_deit_tiny_patch16_224.pth --reprob 0 --opt sgd --data-set CIFAR10 --dist-eval --input-size 224 --lr 1e-2 --weight-decay 1e-4 --epochs 150

我们对不同bsz的训练仅改变了bsz的大小，我们使用的代码是当前最新的。

**Comment by wkcn** [comment]
@Weigaa 
有可能是因为增大batch size后，共享内存不足导致的。
可以在训练的时候输入`df -lh /dev/shm` 查看共享内存占用情况。

**Comment by Weigaa** [comment]
@wkcn感谢您的回答，为了验证这个问题我们分别在bsz为32和512时，同时使用free命令和df -lh /dev/shm命令查看内存使用情况，bsz为32时，程序运行良好，运行1个epoch（1次training+1次test）时检测到的剩余内存和共享内存情况为（纵坐标单位是字节(B)）：
![1653196214](https://user-images.githubusercontent.com/38322329/169679795-aed2b7c2-0a4d-4169-bbc6-85385a178c9b.png)
而在bsz为512时，程序不出意外的出错了，我们观察到的结果是可用内存空间不足（虽然此时共享内存dev/shm仍未占满），如您所说，应当是CPU DRAM内存容量成为了程序瓶颈。
![1653196266(1)](https://user-images.githubusercontent.com/38322329/169679818-970ed0cf-8f4e-4266-8ba0-7117da3765d3.png)
为了解决这个问题（将CPU memory wall问题转移到GPU memory wall问题，让GPU称为限制程序运行的原因），并同时保证程序运行的一致性，我没有减少dataloader的worker用量而是增大了系统的swap空间，这可能会带来一定的速度损失，但是无需修改源代码逻辑。
我不确定使用swap和设置-num_workers 5哪个对速度影响更大，也不确定修改num_workers会不会让精度受到损失，我会在后续尝试并在这个帖子里补充实验结果，以给可能遇到和我一样问题的朋友们提供解决方案。

**Comment by wkcn** [comment]
@Weigaa 感谢你详细的分析！

Batch Size是512的时候，是在第一个epoch的测试阶段内存不足的吗？

关于swap和设置num_workers对速度的影响:
可以打印读取数据的时间

修改engine.py中的train_one_epoch函数
```
data_tic = time.time()
for samples, targets in metric_logger.log_every(data_loader, print_freq, header):
  data_cost = time.time() - data_tic  # 读取数据的时间
  ...
  data_tic = time.time()
```

修改num_workers应该不会影响模型的精度。

**Comment by wkcn** [comment]
Close it. Feel free to reopen it if any problem exists. : )
